higher-order-scaffolders.md

```diff
@@ -0,0 +1,62 @@
+# Higher Order Scaffolders
+
+Start Date: 2020-08-16
+
+Author: @travi
+
+## Summary
+
+Define a pattern for scaffolders that can follow a more rigid process than
+simple plugins, like package or application scaffolders. Because they solve a
+well defined problem, they are able to define answers to the generic questions
+provided by default in the [project-scaffolder](https://github.com/travi/project-scaffolder)
+or language scaffolders. These also need to define details commonly returned as
+results from the plugin/sub-scaffolders, like `dependencies`,
+`devDependencies`, `scripts`, and others.
+
+## Motivation
+
+* Further automate well-defined scenarios that have become repetitive
+* Enable orchestrating project scaffolding tasks at a higher level so that
+  larger tasks can be automated, like creating the initial projects to
+  bootstrap a new organization.
+
+## Detailed Explanation
+
+{{Describe the expected changes in detail, }}
+
+## Rationale and Alternatives
+
+{{Discuss 2-3 different alternative solutions that were considered. This is
+required, even if it seems like a stretch. Then explain why this is the best
+choice out of available ones.}}
+
+* Continue to build package/application plugins and manually answer prompts,
```

Comment:
could you describe if it's possible to just update/redesign the current scaffolders to conform to a strict interface that these can conform to? so in-effect it would end up not caring about if it was a child/parent scaffolder. 

it makes me think a lot of the step-flow framework i worked on at Gain. one of the core design constraints was that each "unit" had to conform to a specific interface, then when you execute them it didn't matter if a given chunk technically had child chunks or not.

hopefully valuable to think about and not just purely noob question, though admittedly i'm having a hard time tracking down/aggregating the shape expectations for the current project and language scaffolders.

Reply:
the plugins already follow a pretty consistent interface, at least for what they are able to return as results. the arguments passed to the plugins or to the project-scaffolder or language scaffolders would be more difficult because they need different information depending on what they are expected to accomplish. there certainly are some common arguments, but also some that are pretty specific.

would i be accurate using the [chain of responsibility pattern](https://refactoring.guru/design-patterns/chain-of-responsibility) to refer to what you are describing? it seems worth at least listing here as an alternative. i think it would be difficult to leverage in this context, for the reasons i just mentioned above. i cant say it cant be a fit, but it would not be a small undertaking to get there. if we were to decide it was a fit, it seems like one taht we would either want to refactor to later or do that separately first and then revisit this concept once that is there. i dont think it would be appropriate to combine that effort with the one that this RFC is for.

Reply:
not quite chain-of-responsibility, but [applying enhancers](https://github.com/form8ion/lift/blob/master/src/enhancers.js) in `lift` is a little closer to along these lines, by design. my intent there was to decouple the actual enhancers from the context where they were used.

Reply:
that's not quite what i had in mind, but i think is honestly better and more all-encompassing 👍 . really interesting. the piece i've been struggling with the most in getting familiar with the codebase is just knowing what to expect where, so shapes going in and shapes coming out. I've seen that we do verification with joi, but i haven't been able to use that to sorta pin that shape to a "plugin" in my head. i'm still at a point where i'm in the trees and having a hard time putting the concepts together into the forest. i think something like this design pattern could go a long way with that.

fair point in regard to when the right time would be for that. seems like a pretty involved change when you can probably make this work as-is right now.